pep-0692.rst

```diff
@@ -0,0 +1,344 @@
+=========
+
+PEP 589 introduced the ``TypedDict`` type constructor that supports dictionary types consisting of string keys and values of potentially different types. This makes ``TypedDict`` a perfect candidate to be used for more precise ``**kwargs`` typing. However, specifing ``**kwargs`` type using a ``TypedDict`` means, as mentioned earlier, that each keyword argument specified by ``**kwargs`` is a ``TypedDict`` itself. For instance:
+
+::
+
+    class Movie(TypedDict):
+        name: str
+        year: int
+    
+    def foo(**kwargs: Movie) -> None: ...
+
+means that each keyword argument in ``foo`` is itself a ``Movie`` dictionary that has a ``name`` key with a string type value and a ``year`` key with an integer type value. Therefore, in order to support specifing ``kwargs`` type as a ``TypedDict`` without breaking current behaviour, a new syntax has to be introduced.
+
+Specification
+=============
+
+To support the aforementioned use case we propose to use the double asterisk syntax inside of the type annotation. The required grammar change is discussed in more detail in section `Grammar Changes`_. Continuing the previous example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+would mean that the ``**kwargs`` comprise two keyword arguments specified by ``Movie`` (i.e. a ``name`` keyword of type ``str`` and a ``year`` keyword of type ``int``). Then, inside the function itself, appropriate type checking would take place:
+
+::
+
+    def foo(**kwargs: **Movie) -> None:
+        kwargs["name"].capitalize()      # OK!
+        kwargs["year"] + 42              # OK!
+        kwargs["name"] + kwargs["year"]  # WRONG! Unsupported operand types for + ("str" and "int")
+
+The ``kwargs`` parameter inside the function should be treated exactly the same way as any other ``TypedDict``.
+
+Function calls with standard dictionaries
+-----------------------------------------
+
+Calling a function that has ``**kwargs`` typed using the ``**kwargs: **Movie`` sytnax with a standard dictionary must raise an error. For example:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie: dict[str, str] = {"name": "Life of Brian", "year": 1979}
+    foo(**movie)  # WRONG! Movie is of type dict[str, str]
+
+    typedMovie: Movie = {"name": "The Meaning of Life", "year": 1983}
+    foo(**typedMovie)  # OK!
+
+
+However, if the dictionary is completely untyped it is up to the type checker whether to raise an error or not:
+
+::
+
+    def foo(**kwargs: **Movie) -> None: ...
+
+    movie = {"name": "Life of Brian", "year": 1979} 
+
+    foo(**movie)  # Error/Success dependant on the type checker.
+
+
+Keyword collisions
+------------------
+
+A ``TypedDict`` that is used to type ``**kwargs`` could potentially contain keys that are already defined in the function's signature. If the duplicate name is a standard argument, an error should be raised. If the duplicate name is a positional only argument, no errors should be raised. For example:
```

Comment:
"an error should be raised" / "no errors should be raised" -- again, unspecified

Reply:
All of this should be type checking-only. Maybe just put a note at the beginning of the spec saying that the new annotation has no runtime effect.

Reply:
As mentioned above, I'd avoid using the term "raise" when talking about errors emitted by a type checker. Use "generated" or "emitted" or "reported" instead. The term "raise" implies a runtime exception.